odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Note that this is the same as the gradient of the original functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return self.left.gradient
+
+    def proximal(self, sigma=1.0):
+        """Proximal operator of the FunctionalScalarSum.
+
+        This is the same as the proximal operator of the original
+        functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+        """
+        return self.left.proximal(sigma)
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of FunctionalScalarSum."""
+        return self.left.conjugate_functional - self.scalar
+
+    def derivative(self, point):
+        """Returns the derivative operator of FunctionalScalarSum.
+
+        Parameters
+        ----------
+        point : `LinearSpaceVector`
+            The point in which the gradient is evaluated.
+
+        Returns
+        -------
+        out : `DerivativeOperator`
+            The linear operator that maps ``x --> <x, grad_f(point)>``.
+        """
+        return self.left.derivative(point)
+
+
+class TranslatedFunctional(Functional):
+
+    """Implementation of the translated functional.
+
+    Given a functional ``f`` and an element ``translation`` in the domain of
+    ``f``, this corresponds to the functional ``f(. - translation)``.
+    """
+
+    def __init__(self, func, translation):
+        """Initialize a new instance.
+
+        Given a functional ``f(.)`` and a vector ``translation`` in the domain
+        of ``f``, this corresponds to the functional ``f(. - y)``.
+
+        Parameters
+        ----------
+        func : `Functional`
+            Functional which is to be translated.
+        translation : `LinearSpaceVector`
+            Element in ``func.domain``, with which the argument is translated.
+        """
+
+        if not isinstance(func, Functional):
+            raise TypeError('`func` {!r} not a `Functional` instance'
+                            ''.format(func))
+
+        if translation not in func.domain:
+            raise TypeError('`translation` {!r} not in func.domain {!r}'
+                            ''.format(translation.space, func.domain))
+
+        super().__init__(domain=func.domain, linear=False,
+                         grad_lipschitz=func.grad_lipschitz)
+
+    # TODO: Add case if we have translation -> scaling -> translation?
```

Comment:
Could be something to think of. Maybe we don't need to do it right away. In any case, indent the comment :-)